### PR TITLE
add gsAdditiveOp

### PR DESCRIPTION
**Pull request by stefantakacs**
Allows to construct additive Schwarz type smoothers.

This will be required for multipatch multigrid stuff

### CODE REVIEW

File: src/gsSolver/gsAdditiveOp.h [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+#include <gsSolver/gsPreconditioner.h>
+#include <gsSolver/gsMatrixOp.h>
+#include <gsCore/gsMultiPatch.h>
+
+namespace gismo
+{
+
+/// @brief Generic preconditioner which applies an arbitrary linear operator to the residual.
+///
+/// \code{.cpp}
```
Comment (filiatra) [inline src/gsSolver/gsAdditiveOp.h]: would be nice to add as documentation the mathematical formula for completeness

File: src/gsSolver/gsAdditiveOp.h [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+///          )
+///       );
+/// \endcode
+///
+/// but much faster.
+///
+/// @ingroup Solvers
+
+template<class T>
+class gsAdditiveOp : public gsLinearOperator<T>
```
Comment (filiatra) [inline src/gsSolver/gsAdditiveOp.h]: could be  GISMO_FINAL

File: src/gsSolver/gsAdditiveOp.h [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    /// Shared pointer
+    typedef memory::shared_ptr<gsAdditiveOp> Ptr;
+
+    /// Unique pointer
+    typedef memory::unique_ptr<gsAdditiveOp> uPtr;
+
+    /// Default Constructor
+    gsAdditiveOp() : m_transfers(), m_ops() {}
+
+    /// Constructor
```
Comment (filiatra) [inline src/gsSolver/gsAdditiveOp.h]: Please document each argument (same for make(..) and os on)
eg. what is the size of the m_transfers that is expected ?

File: src/gsSolver/gsAdditiveOp.h [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    index_t cols() const
+    {
+        GISMO_ASSERT( !m_transfers.empty(), "gsAdditiveOp::cols does not work for 0 operators." );
+        return m_transfers[0].rows();
+    }
+
+protected:
+    TransferContainer m_transfers;
```
Comment (filiatra) [inline src/gsSolver/gsAdditiveOp.h]: Please add short domunetation for the members

File: src/gsSolver/gsAdditiveOp.h [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+    Author(s): S. Takacs
+*/
+
+#pragma once
+
+#include <gsSolver/gsPreconditioner.h>
+#include <gsSolver/gsMatrixOp.h>
+#include <gsCore/gsMultiPatch.h>
```
Comment (filiatra) [inline src/gsSolver/gsAdditiveOp.h]: why is gsMultiPatch included? is it used ?

File: src/gsSolver/gsAdditiveOp.h [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+    Author(s): S. Takacs
+*/
+
+#pragma once
+
+#include <gsSolver/gsPreconditioner.h>
+#include <gsSolver/gsMatrixOp.h>
```
Comment (filiatra) [inline src/gsSolver/gsAdditiveOp.h]: seems that gsMatrixOp.h is only used in hpp, should be removed ?

File: src/gsSolver/gsAdditiveOp.h [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    This Source Code Form is subject to the terms of the Mozilla Public
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+    Author(s): S. Takacs
+*/
+
+#pragma once
+
+#include <gsSolver/gsPreconditioner.h>
```
Comment (filiatra) [inline src/gsSolver/gsAdditiveOp.h]: seems gsPreconditioner is not used by this class ?

**Comment by stefantakacs** [comment]
@filiatra  I have also added a few GISMO_FINAL to the other classes derived from gsLinearOperator that should be final.

**Comment by stefantakacs** [comment]
@filiatra  jarlesog and me were wondering if it is possible to withdraw a review.

**Pull request merged**